### PR TITLE
JS: skip binary files when extracting JavaScript

### PR DESCRIPTION
**Pull request by erik-krogh**

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java [hunk 1]
```diff
@@ -90,6 +155,29 @@ protected boolean contains(File f, String lcExt, ExtractorConfig config) {
...
+      /** Computes if `f` is a binary file based on whether the initial `fileHeaderSize` bytes are printable UTF-8 chars. */
+      private boolean isBinaryFile(File f, String lcExt, ExtractorConfig config) {
+        try (FileInputStream fis = new FileInputStream(f)) {
+          byte[] bytes = new byte[fileHeaderSize];
+          int length = fis.read(bytes);
+
+          if (length == -1) return false;
+
+          // Avoid invalid or unprintable UTF-8 files.
+          if (config.getDefaultEncoding().equals(UTF8_CHARSET.name()) && hasUnprintableUtf8(bytes, length)) {
```
Comment (esbena) [inline javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java:171]: So if `config.getDefaultEncoding().equals(UTF8_CHARSET.name()) == false`, then this method always returns `false` without reading any bytes. Can we move that check outside the `try`-block?

File: javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java [hunk 2]
```diff
@@ -42,7 +42,68 @@
       Pattern.compile("^(?s)\\s*\\{\\s*\"([^\"]|\\\\.)*\"\\s*:.*");
 
   /** The charset for decoding UTF-8 strings. */
-  private static final Charset UTF8_CHARSET = Charset.forName("UTF-8");
+  private static final Charset UTF8_CHARSET = StandardCharsets.UTF_8;
```
Comment (esbena) [inline javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java:45]: Hmm. Wouldn't it be cleaner just to inline `UTF8_CHARSET` now that it isn't a method call any more?

File: javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java [hunk 3]
```diff
@@ -42,7 +42,68 @@
...
+
+  /**
+   * Returns true if the byte sequence contains invalid UTF-8 or unprintable ASCII characters.
+   */
+  private static boolean hasUnprintableUtf8(byte[] bytes, int length) {
```
Comment (esbena) [inline javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java:50]: I trust that this is a verbatim move.

File: javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java [hunk 4]
```diff
@@ -42,7 +42,68 @@
...
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  /** Returns the index after the initial BOM, if any, otherwise 0. */
+  private static int skipBOM(byte[] bytes, int length) {
```
Comment (esbena) [inline javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java:98]: I trust that this is a verbatim move.

**Review by esbena — COMMENTED** [review]
LGTM. I have to minor cleanup suggestions.

Again, this requires an extractor version string update.

File: javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java (hunk 3 above)
Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/extractor/FileExtractor.java:50]: > I trust that this is a verbatim move.

It is, byte for byte - except for adjusted indentation.

**Review by esbena — APPROVED** [review]

**Pull request merged**